src/umap.cpp

```diff
@@ -403,12 +503,163 @@ void _umap::pagefault_event(const struct uffd_msg& msg)
+            if (msg.arg.pagefault.flags & (UFFD_PAGEFAULT_FLAG_WP | UFFD_PAGEFAULT_FLAG_WRITE)) {
+                pages_in_memory[bufidx].mark_page_dirty();
+                disable_wp_on_pages((uint64_t)page_begin, 1);
+            }
+
+            struct uffdio_range wake;
+            wake.start = (uint64_t)page_begin;
+            wake.len = page_size; 
+
+            if (ioctl(userfault_fd, UFFDIO_WAKE, &wake) == -1) {
+                perror("ioctl(UFFDIO_WAKE)");
+                exit(1);
+            }
+            continue;
+        }
+
+        //
+        // Page not in memory, read it in and evict someone
+        //
+	int file_id=0;
+	off_t offset=(uint64_t)page_begin - (uint64_t)segment_address;
+        //find the file id and offset number                                              
+        while (offset>=frame_size)
+	{
+            file_id++;
+            offset-=frame_size;
+	}
+
+        ssize_t pread_ret = pread(fd_list[file_id], tmppagebuf, page_size,
+                       offset+fits_offset);
+
+        if (pread_ret == -1) {
+            perror("pread failed");
+            exit(1);
+        }
+
+        if (pages_in_memory[next_page_alloc_index].get_page()) {
+            delete_page_index(pages_in_memory[next_page_alloc_index].get_page());
+            evict_page(pages_in_memory[next_page_alloc_index]);
+        }
+
+        pages_in_memory[next_page_alloc_index].set_page(page_begin);
+        add_page_index(next_page_alloc_index, page_begin);
+
+        if (msg.arg.pagefault.flags & (UFFD_PAGEFAULT_FLAG_WP | UFFD_PAGEFAULT_FLAG_WRITE)) {
+            disable_wp_on_pages((uint64_t)page_begin, 1);
+            pages_in_memory[next_page_alloc_index].mark_page_dirty();
+        }
+        else {
+            pages_in_memory[next_page_alloc_index].mark_page_clean();
+        }
+
+        next_page_alloc_index = (next_page_alloc_index +1) % page_buffer_size;
+
+        struct uffdio_copy copy;
+        copy.src = (uint64_t)tmppagebuf;
+        copy.dst = (uint64_t)page_begin;
+        copy.len = page_size;
+
+        if (msg.arg.pagefault.flags & (UFFD_PAGEFAULT_FLAG_WP | UFFD_PAGEFAULT_FLAG_WRITE)) {
+            copy.mode = 0;
+            if (ioctl(userfault_fd, UFFDIO_COPY, &copy) == -1) {
+                perror("ioctl(UFFDIO_COPY wake)");
+                exit(1);
+            }
+        }
+        else {
+            copy.mode = UFFDIO_COPY_MODE_DONTWAKE;
+            if (ioctl(userfault_fd, UFFDIO_COPY, &copy) == -1) {
+                perror("ioctl(UFFDIO_COPY nowake)");
+                exit(1);
+            }
+            enable_wp_on_pages_and_wake((uint64_t)page_begin, 1);
+        }
+    }
+    next_page_alloc_index = (next_page_alloc_index +1) % page_buffer_size;
+}
 
 void _umap::evict_page(umap_page& pb)
 {
     uint64_t* page = (uint64_t*)pb.get_page();
 
-    if (pb.page_is_dirty()) {
+    if (pb.page_is_dirty()&&(backingfile_fd!=-1)) {//--------if multi-file, no pwrite needed
```

Comment:
I am not sure I understand why we have the additional check of backingfile_fd.  The page is marked as clean when it is allocated and only becomes dirty if someone writes to the memory.  If we want to insure that the memory is truly read-only, we should handle that in the handler when/if we handle a write-fault.